src/QueryManager.ts

```diff
@@ -0,0 +1,222 @@
+import Store from '@orbit/store'
+
+import {
+  Term,
+  OngoingQueries,
+  RecordObject,
+  Expressions,
+  Queries,
+  Subscriptions,
+  QueryRefs,
+  Statuses,
+  QueryCacheOptions,
+} from './types'
+import { Transform, RecordOperation, RecordIdentity } from '@orbit/data'
+import { identityIsEqual } from './helpers'
+
+export class QueryManager<E extends { [key: string]: any } = any>  {
+  _extensions: E
+  _store: Store
+  _ongoingQueries: OngoingQueries
+
+  subscriptions: Subscriptions
+  statuses: Statuses
+
+  constructor (orbitStore: Store, extensions?: E) {
+    this._extensions = extensions || {} as E
+    this._store = orbitStore
+    this._ongoingQueries = {}
+
+    this.subscriptions = {}
+    this.statuses = {}
+  }
+
+  generateQueryRef (queries: Queries) {
+    const terms = this._extractTerms(queries)
+
+    let queryRef = JSON.stringify(terms)
+
+    if (this.subscriptions[queryRef]) return queryRef
```

Comment:
I favour single return points in functions when possible (and I don't really like a single line if statements), e.g.:

```
if (!this.subscriptions[queryRef]) {
  this.subscriptions[queryRef] = { listeners: 0, terms }
}

return queryRef
```